### PR TITLE
Fix Overlap Tests in Biochemistry Incarnations

### PR DESCRIPTION
**Pull request by giuluck**
Fixed test testNoOverlapInSimulation1 in TestBioRect2DEnvironmentNoOverlap.java, with minor refactoring of other tests due to internal changes.

### CODE REVIEW

File: alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java [hunk 1]
```diff
@@ -714,39 +698,37 @@ public void finished(final Environment<Double, Euclidean2DPosition> env, final T
...
-                            System.out.println("nodes: ");
-                            System.out.println("center : " + env.getPosition(n));
-                            listOverlapping.forEach(c -> System.out.println("In range : " + env.getPosition(c)));
-                            listOverlapping.forEach(c -> System.out.println("distance : " + env.getPosition(c).getDistanceTo(env.getPosition(n))));
-                        }
-                    });
-                    return posResult;
-                }
+                        .map(n -> (CellWithCircularArea<Euclidean2DPosition>) n)
+                        .flatMap(n -> env.getNodesWithinRange(n, n.getDiameter() - DELTA).stream().peek(c -> {
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: Peeking on a stream is very, very non functional. Try to collect the information, and fail only on termination:
```java
n -> env.getNodesWithinRange(n, n.getDiameter() - DELTA).stream()
    .map(n -> new Pair(c, n))
```

File: alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java [hunk 2]
```diff
@@ -714,39 +698,37 @@ public void finished(final Environment<Double, Euclidean2DPosition> env, final T
...
+                            fail("Nodes " + n.getId() + env.getPosition(n) + " and " + c.getId() + env.getPosition(c) +
+                                    " are overlapping. Their distance is: " + env.getDistanceBetweenNodes(n, c));
+                        }))
+                        .findAny()
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: With the proposed change, you can here do something like:
```java
.findAny()
.ifPresent(it -> fail(...))
```

File: alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java [hunk 3]
```diff
@@ -39,12 +31,14 @@
...
 
+import static org.junit.Assert.*;
+
 /**
  *
  */
 public class TestBioRect2DEnvironmentNoOverlap {
 
-    private static final double MAX_PRECISION = 0.9999999999;
+    private static final double DELTA = 1e-6;
```
Comment (DanySK) [inline alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: Looks rather arbitrary: up to which value does the test work? What happens if you put a zero? What if you put `nextUp(0)`?

Comment (giuluck) [reply alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: The test works for values up to DELTA=1e-7 and starts failing with DELTA=1e-8 (same for DELTA=0, obviously, but giving in the error log that two nodes with distance d=1.0 are overlapping, which should be acceptable so a DELTA is needed).
Eventually, as nextUp(0) returns a value which is about 1e-45, the test wil fail too.

File: alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java [hunk 4]
```diff
@@ -782,7 +764,7 @@ private boolean thereIsOverlap(final Environment<Double, Euclidean2DPosition> en
                                 .filter(c -> env.getDistanceBetweenNodes(c, n)
                                         < ((((CellWithCircularArea<Euclidean2DPosition>) n).getRadius()
```
[the remark this answers is not in the document]
Comment (giuluck) [reply alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: This was an old piece of code I didn't change.
In the last commit, I made a general refactoring of those two tests removing code duplications and fixing the code style.

File: alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java (hunk 3 above)
Comment (DanySK) [reply alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: I'd say here documenting (with a comment in code) which is the limit at which the test works is fine.

Comment (giuluck) [reply alchemist/alchemist-incarnation-biochemistry/src/test/java/it/unibo/alchemist/test/TestBioRect2DEnvironmentNoOverlap.java]: Now it should be ok

**Pull request merged**